[AutoDiff] Fixes memory leaks in autodiff linear map context allocation builtins

jkshtj opened the pull request:
When the differentiating a function containing loops, we allocate a linear map context object on the heap. This context object may store non-trivial objects, such as closures,that need to be freed explicitly, at program exit. This PR fixes the autodiff linear map context allocation builtins to destroy and deallocate any such objects.

Fixes #67323

jkshtj committed:
Fixes memory leaks in autodiff linear map context allocation builtins

Fixes #67323

jkshtj commented:
@asl @BradLarson Could I get a review on this?

BradLarson commented:
@swift-ci Please test.

asl commented on include/swift/Runtime/RuntimeFunctions.def:
```diff
@@ -2273,12 +2273,12 @@ FUNCTION(TaskGroupDestroy,
          EFFECT(Concurrency))
 
-// AutoDiffLinearMapContext *swift_autoDiffCreateLinearMapContext(size_t);
+// AutoDiffLinearMapContext *swift_autoDiffCreateLinearMapContext(const Metadata *);
```
So, there will be no link errors for things compiled with old compiler and linked with new runtime and vice versa?

jkshtj replied:
Ah, it slipped my mind and I meant to ask this. I wasn't able to figure out how to make this a link time error.

asl replied:
The easiest way: rename functions :) Hard way: invent some backward compatibility scheme. @rxwei any suggestions? I think we do not care about ABI stability for now, isn't it?

jkshtj replied:
Haven't made a decision on this in the latest revision. Waiting on @rxwei 's input. If breaking ABI is going to be a concern we could deprecate existing builtins instead of completely removing them.

jkshtj replied:
Should we deprecate the old ones as well? They do lead to memory leaks.

rxwei replied:
What do you mean by deprecate? We just don't call it

rxwei replied:
I mean, this is not an API. There is not a real way to deprecate it.

asl replied:
@jkshtj I believe what @rxwei means is as follows:
 - Add new functions, say `swift_autoDiffCreateLinearMapContext2` etc. with the desired functionality
 - Keep old functions as-is

As a result: old code would still be able to run with the new runtime. The behavior will be the same, well, it will leak. The new code will call that new runtime functions and will be leak-less.

asl commented on test/AutoDiff/SILGen/autodiff_builtins.swift:
```diff
@@ -84,24 +84,19 @@ func applyDerivative_f1_vjp<T: AdditiveArithmetic & Differentiable>(t0: T) -> (T
-// CHECK:   [[BUF:%.*]] = builtin "autoDiffAllocateSubcontext"([[BORROWED_CTX]] : $Builtin.NativeObject, {{.*}} : $Builtin.Word) : $Builtin.RawPointer
+// CHECK-LABEL: sil{{.*}}@test_context_builtins : $@convention(thin) <T> (@in_guaranteed T) -> () {
+// CHECK: bb0({{%.*}} : $*T):
+// CHECK:   [[CTX:%.*]] = builtin "autoDiffCreateLinearMapContext"<T>({{%.*}} : $@thick T.Type) : $Builtin.NativeObject // users: {{.*}}
```
Is it really necessary to make these builtins generic?

jkshtj replied:
So the actual builtins are actually not generic. I'm not entirely sure why, but I simply changed the `size_t` arg to a metatype and this is the code that the compiler is generating. Even the SIL that's generated from source does not have generic builtins.

jkshtj replied:
On a side note, I had a hard time figuring out how to get a thick metatype for a non-class type, which is why I had to make the test function generic. Would you know if it's possible to get a thick metatype for a struct/enum here?

asl replied:
Ah, sorry, I misread the test changes. It's the test function `test_context_builtins` that is now generic!

asl commented on test/AutoDiff/stdlib/callee_differential_not_leaked_in_func_with_loops.swift:
```diff
@@ -0,0 +1,61 @@
+
+import _Differentiation
+import StdlibUnittest
+import CoreFoundation
```
I do not think this will work on Linux / Windows – no CoreFoundation there. Can we use the existing `Tracked` wrappers btw?

jkshtj replied:
Ah sorry, that CoreFoundation import is unused from. I was trying something out. I'll remove it.

asl commented on test/AutoDiff/stdlib/callee_differential_not_leaked_in_func_with_loops.swift:
```diff
@@ -0,0 +1,61 @@
+// that need to be freed explicitly, at program exit. This test verifies that the 
+// autodiff runtime destroys and deallocates any such objects.
+
+extension LifetimeTracked: AdditiveArithmetic {
```
Can we use the existing Tracked wrappers btw?

jkshtj replied:
See comment above.

jkshtj committed:
Reintegrates BumpPtrAllocator in AutoDiffLinearMapContext

This commit builds on top of a previous commit that fixed memory leak
issues in the autodiff loop context allocation builtins. The previous
commit allocated loop contexts in a box, which is known to have memory
scaling issues. This commit modifies AutoDiffLinearMapContext to perform
memory allocation using the BumpPtrAllocator, like it used to, while
introducing new bookkeeping logic that enables AutoDiffLinearMapContext
to prevent memory leaks, unlike before.

jkshtj commented:
@asl as per offline discussion w/ @BradLarson, since we're already aware of the [memory scaling issues](https://github.com/apple/swift/pull/34886) with allocating loop context objects directly on the heap I have created a new revision to reintegrate `BumpPtrAllocator` in `AutodiffLinearMapContext` while still preventing memory leaks.

asl commented:
> since we're already aware of the https://github.com/apple/swift/pull/34886 with allocating loop context objects directly on the heap
@jkshtj Well, this was quite different case, as the chain of boxes were explicitly created ;) Anyway, as we already discussed, as soon as you had the initial implementation on top of the boxes, it was possible to change the underlying implementation to something more lightweight, as we certainly did not want to reference count the pullback tuples.

asl commented:
@swift-ci please test

jkshtj committed:
Creates new autodiff loop context allocation builtins and preserves old ones

This commit builds on top of a previous commit that fixed memory leak
issues in the autodiff loop context allocation builtins. The previous
commit changed the existing builtins. However, inorder to maintain ABI
compatibility and not break other consumers of autodiff, this commit
creates newly named builtins that don't leak memory and preserves the
older builtins.

As a result, now, code compiled with an older frontend and using a newer
stdlib will continue to work. But code compiled with a newer frontend
and using older stdlib will have undefined behavior.

BradLarson commented:
@swift-ci Please test

asl commented:
@swift-ci please test windows

asl commented on include/swift/AST/Builtins.def:
```diff
@@ -994,6 +994,12 @@ BUILTIN_MISC_OPERATION_WITH_SILGEN(AutoDiffProjectTopLevelSubcontext, "autoDiffP
 // autoDiffAllocateSubcontext: (Builtin.NativeObject, Builtin.Word) -> Builtin.RawPointer
 BUILTIN_MISC_OPERATION_WITH_SILGEN(AutoDiffAllocateSubcontext, "autoDiffAllocateSubcontext", "", Special)
 
+// autoDiffCreateLinearMapContext2: (T.Type) -> Builtin.NativeObject
```
Some bike-shedding: maybe rename these to `autoDiffCreateLinearMapContextWithType` etc.?

asl commented on lib/AST/Builtins.cpp:
```diff
@@ -1608,8 +1608,8 @@ static ValueDecl *getBuildComplexEqualitySerialExecutorRef(ASTContext &ctx,
 
 static ValueDecl *getAutoDiffCreateLinearMapContext(ASTContext &ctx,
                                                     Identifier id) {
-  return getBuiltinFunction(
```
Remove formatting change?

jkshtj replied:
Removed

asl commented on lib/IRGen/GenCall.cpp:
```diff
@@ -5479,8 +5479,9 @@ IRGenFunction::getFunctionPointerForResumeIntrinsic(llvm::Value *resume) {
   return fnPtr;
 }
 
-Address irgen::emitAutoDiffCreateLinearMapContext(
```
ditto

jkshtj replied:
Removed

asl commented on lib/IRGen/GenCall.cpp:
```diff
@@ -5500,8 +5501,9 @@ Address irgen::emitAutoDiffProjectTopLevelSubcontext(
   return Address(call, IGF.IGM.Int8Ty, IGF.IGM.getPointerAlignment());
 }
 
-Address irgen::emitAutoDiffAllocateSubcontext(
```
ditto

jkshtj replied:
Removed

asl commented on lib/IRGen/GenCall.cpp:
```diff
@@ -5500,8 +5501,9 @@ Address irgen::emitAutoDiffProjectTopLevelSubcontext(
 
-Address irgen::emitAutoDiffAllocateSubcontext(
-    IRGenFunction &IGF, Address context, llvm::Value *size) {
+Address irgen::emitAutoDiffAllocateSubcontext(IRGenFunction &IGF,
```
Do we need to keep these functions after all? We are not going to call these builtins from the compiler.

jkshtj replied:
Removed

asl commented on lib/AST/Builtins.cpp:
```diff
@@ -1625,6 +1625,20 @@ static ValueDecl *getAutoDiffAllocateSubcontext(ASTContext &ctx,
       ctx.TheRawPointerType);
 }
 
+static ValueDecl *getAutoDiffCreateLinearMapContext2(ASTContext &ctx,
```
As the functions above with old builtins are never called from the compiler, I'd steal their names for the helpers that call new builtins

jkshtj replied:
I think it'll be better to keep the helpers symmetric with the names of the builtins. Otherwise `autoDiffProjectTopLevelSubcontext` will be the odd one out in different places.

BradLarson commented:
@swift-ci Please test

asl reviewed (approved):
Thanks!
